test/testutilscompare.cpp

```diff
@@ -0,0 +1,184 @@
+        {u"ABC100"_qs, u"ABC99"_qs, CompareResult::Greater, CompareResult::Greater},
+
+        {u"100abc"_qs, u"99abc"_qs, CompareResult::Greater, CompareResult::Greater},
+        {u"100ABC"_qs, u"99abc"_qs, CompareResult::Greater, CompareResult::Greater},
+        {u"100abc"_qs, u"99ABC"_qs, CompareResult::Greater, CompareResult::Greater},
+        {u"100ABC"_qs, u"99ABC"_qs, CompareResult::Greater, CompareResult::Greater},
+
+        {u"😀😀😀99"_qs, u"😀😀😀100"_qs, CompareResult::Less, CompareResult::Less},
+        {u"😀😀😀100"_qs, u"😀😀😀99"_qs, CompareResult::Greater, CompareResult::Greater}
+    };
+
+    void testCompare(const TestData &data, const int actual, const CompareResult expected)
+    {
+        const auto errorMessage = u"Wrong result. LHS: \"%1\". RHS: \"%2\". Result: %3"_qs
+            .arg(data.lhs, data.rhs, QString::number(actual));
+
+        switch (expected)
+        {
+        case CompareResult::Equal:
+            QVERIFY2((actual == 0), qPrintable(errorMessage));
+            break;
+        case CompareResult::Greater:
+            QVERIFY2((actual > 0), qPrintable(errorMessage));
+            break;
+        case CompareResult::Less:
+            QVERIFY2((actual < 0), qPrintable(errorMessage));
+            break;
+        default:
+            QFAIL("Unhandled case");
+            break;
+        }
+    }
+
+    void testLessThan(const TestData &data, const bool actual, const CompareResult expected)
+    {
+        const auto errorMessage = u"Wrong result. LHS: \"%1\". RHS: \"%2\". Result: %3"_qs
+            .arg(data.lhs, data.rhs, QString::number(actual));
+
+        switch (expected)
+        {
+        case CompareResult::Equal:
+        case CompareResult::Greater:
+            QVERIFY2(!actual, qPrintable(errorMessage));
+            break;
+        case CompareResult::Less:
+            QVERIFY2(actual, qPrintable(errorMessage));
+            break;
+        default:
+            QFAIL("Unhandled case");
+            break;
+        }
+    }
+}
+#endif
+
+class TestUtilsCompare final : public QObject
+{
+    Q_OBJECT
+    Q_DISABLE_COPY_MOVE(TestUtilsCompare)
+
+public:
+    TestUtilsCompare() = default;
+
+#ifndef QBT_USE_QCOLLATOR  // only test qbt own implementation, not QCollator
+private slots:
+    void testNaturalCompareCaseInsensitive() const
+    {
+        const Utils::Compare::NaturalCompare<Qt::CaseInsensitive> cmp;
+
+        for (const TestData &data : testData)
+            testCompare(data, cmp(data.lhs, data.rhs), data.caseInsensitiveResult);
+    }
+
+    void testNaturalCompareCaseSensitive() const
+    {
+        const Utils::Compare::NaturalCompare<Qt::CaseSensitive> cmp;
+
+        for (const TestData &data : testData)
+            testCompare(data, cmp(data.lhs, data.rhs), data.caseSensitiveResult);
+    }
+
+    void testNaturalLessThanCaseInsensitive() const
+    {
+        const Utils::Compare::NaturalLessThan<Qt::CaseInsensitive> cmp;
```

Comment:
There are several warnings like this one.

Reply:
I see it in GitHub "Files changed" tab:
> 'cmp': 'const' automatic data initialized with compiler generated default constructor produces unreliable results

Reply:
I suppressed it.